### PR TITLE
Get rid of bitcoind_rpc as configuration of `fedimint-wallet`

### PR DESCRIPTION
**Pull request by dpc**
The wallet module already gets the API (`dyn Trait`) injected. It does not need to know by itself where is the bitcoin data comming from. It simplifies a lot of other things, where we don't need to fake things we don't need anyway.

The end application is responsible for providing wallet module with blockchain data: be that bitcoind rpc client, or (in the future) electrum client or something entirely different.

From an operational PoV, the bitcoind rpc port is also so important, that it's best to bring it upfront instead of hidding in a module config.

### CODE REVIEW

**Comment by dpc** [comment]
Please someone double check the `tmuxinator`. It seems to be working but something was odd when I ran it.

**Comment by elsirion** [comment]
This would prevent us from implementing a generic constructor for all modules that consumes type-erased config and returns a type-erased module without additional arguments. It might be worth for being able to inject other bitcoin backends, but we should consider it. Other solutions could be:

* Implement all backends in the wallet module and make them selectable in the config (bitcoin backend enum)
* Only implement creating a bitcoind backend from config but keeping a secondary constructor around that allows injecting a different bitcoin backend

**Comment by dpc** [comment]
> This would prevent us from implementing a generic constructor for all modules that consumes type-erased config and returns a type-erased module without additional arguments. 

In this case we could make `dyn BitcoinRpc` instance a global item that modules can access from some global `struct`. This assumes that many modules might want to talk to Bitcoin Node.

Alternatively - we don't have to have a generic constructor for **all** modules. We could only have it for subset: 3rd party ones. It's a nice property to have, but it is strictly less flexible one, as one can only pass to a module just little bit of text.

Another - more architecturally involving - approach would be to have some "providers" abstraction. Where `provider_registry: BTreeMap<Sting, Provider>` and `struct Provider(Arc<dyn ProviderTrait>)`, where providers are a bit like modules, but their purpose is only of providing access to certain functionality (like calling bitcoind node). Then modules could look up providers they expect during initialization, downcast them to the interface they expect, and use as they see fit. Modules would document providers they need. Yes - it is another complication, but then we don't have to hardcode making functionality of rpc calls to bitcoind and pass connection string to for every module that wants to talk with it.

**Comment by elsirion** [comment]
The provider architecture sounds like a good option later on. I just wanted to avoid us painting ourselves into a corner.

**Review by elsirion — APPROVED** [review]
I'm ok with it (just needs a rebase). Would like to hear what @jkitman thinks.

File: fedimintd/src/ui/mod.rs [hunk 1]
```diff
@@ -110,25 +109,9 @@ async fn deal(
...
-    let user_pass = parts[0].to_string();
-    let btc_rpc_address = parts[1].to_string();
-    let parts: Vec<&str> = user_pass.split(':').collect();
-    let btc_rpc_user = parts[0].to_string();
-    let btc_rpc_pass = parts[1].to_string();
-    let btc_rpc = BitcoindRpcCfg {
-        btc_rpc_address,
-        btc_rpc_user,
-        btc_rpc_pass,
-    };
```
Comment (justinmoon) [inline fedimintd/src/ui/mod.rs:125]: 😬

Comment (elsirion) [reply fedimintd/src/ui/mod.rs:125]: Right, if we want to make it easily user-configurable the "enum in module (or server) cfg" approach might be better.

```rust
enum BitcoinBackend {
    Bitcoind { … /* connect data */ }
    Electrum { … /* connect data */ }
    Esplora { … /* connect data */ }
}
```

You could still create the bitcoin backend outside the module if you want to, but env vars are a hassle to deal with when everything else can be done via the UI.

File: client/client-lib/src/wallet/mod.rs [hunk 2]
```diff
@@ -356,11 +356,6 @@ mod tests {
                 },
                 &ConfigGenParams::new().attach(WalletConfigGenParams {
                     network: bitcoin::network::constants::Network::Regtest,
-                    bitcoin_rpc: BitcoindRpcCfg {
-                        btc_rpc_address: "localhst".to_string(),
-                        btc_rpc_user: "bitcoin".to_string(),
-                        btc_rpc_pass: "bitcoin".to_string(),
```
Comment (justinmoon) [inline client/client-lib/src/wallet/mod.rs:362]: Nice getting rid of hard-coded credentials

**Comment by dpc** [comment]
> You could still create the bitcoin backend outside the module if you want to, but env vars are a hassle to deal with when everything else can be done via the UI.

@justinmoon  @elsirion I think we have a different points of view we look at it. The way I look at it is from devops / sysadmin view. Let me just share it and tell me what you think.

I think it comes from doing SaaS devops kind of things in the recent past, and I'm not sure if it squares with how Fedimint will be used.

If you think about a service in a business SaaS deployment, with some (micro)sevices, the network setup is kind of provided externally. Where the database or a dependency service endpoint is is input data from the outside, and usually have to be glued together by some orchestration system (kubernetes or something). Such configuration details are provided as arguments to the application via command line / environment / config file and not something that would be tweakable from the UI.

From this PoV there's a clear distinction between a system administrator setting it up, and day to day operator (even if it's often the same person/team in practice).

The fact the rpc address to the Bitcoin node is hardcoded somewhere in a user config and requires going to some UI to change is a PITA from that perspective, because it's not that easily "scriptable".

So when I want someone to start  fedimintd as a docker container, ideally I'd like to be able to do:


```
docker run -e FM_BITCOIND_RPC=bitcoind:2342 start
```

or a corresponding thing from a systemd unit file, etc. and provide the Bitcoin host as an argument.

Now I guess, you tend to have a perspective of `fedimintd` being more of a... application/appliance for the end user, where the operator kind of starts it once with minimum amount of input, and then clicks things the UI, including network configuration.

I'm not sure how to reconcile these views. As a Unix pro, I'm biased. I can't help to think of `fedimintd` as a system deamon very much like `nginx`, `posgresql` or something like that.

**Comment by elsirion** [comment]
**tl;dr: ** you are probably right, let's do it this way for now and optimize for medium-technical users later. It will work great for my nix deployment anyway :stuck_out_tongue_winking_eye: 

That line of argument does make sense, it would also imply doing the same for e.g. announce addresses since they are dependent on the reverse proxy/DNS setup and not really fedimint-defined.

Being cloud-ready is nice, but we should not make individual deployments by non-technical users overly burdensome imo. E.g. if you install fedimint as a debian package it should just run and show something useful in the setup UI after you set up the TLS reverse proxy (which hopefully won't be necessary once we get Tor HS support).

# Idea 1

For this specific case of bitcoin rpc we could just assume that bitcoind runs on localhost on the default port with a default cookie file readable by the fedimint user. If not we tell the user in the setup UI to either setup bitcoind in this way or set the appropriate env variables in the systemd file.

# Idea 2

Alternatively we could extend the parsing code for our configs such that any value that's not present in the config file will be read from the environment. E.g. if the local wallet config doesn't contain a bitcoin RPC config, then we'll try to read `FM_WALLET_BTC_RPC` (schema: `FM_<module_upper>_<field-name-upper>`).

This means that for platforms where we control the deployment and can set the env var to something useful the config generator can skip that config field, while on platforms where that might not be possible it's easily user-configurable (e.g. if we ship fedimint as a debian package where we can't include the right env var in the systemd service file because it will depend on the concrete deployment).

Both options might be too much complexity to reduce the still complex setup (reverse proxy, DNS, …) by just a bit. We can always introduce these fixes later on again once this becomes the low hanging fruit for ease-of-setup.

**Comment by dpc** [comment]
> Being cloud-ready is nice, but we should not make individual deployments by non-technical users overly burdensome imo. E.g. if you install fedimint as a debian package it should just run and show something useful in the setup UI after you set up the TLS reverse proxy (which hopefully won't be necessary once we get Tor HS support)

In these situations it is due to distro packaging why even very complex and "low level" software appears easy to install and configure. The problem here is that it requires per-platform effort. One can have nice Ubuntu dpkg builds, nice ArchLinux aur repo, nice Gentoo emerge scripts, nix package, helm chart, kubernetes y8m's... and the list keeps on going. Usually project developers focus on command line / config & documentation, and then power-users build integration around it over time.



BTW. There are two aspects to this discussion worth thinking about separately:

1. How do we provide the configuration settings from the outside (env vars) vs inside (configured in the UX).
2. Injecting interfaces (`dyn BitocinRpc`) vs injecting only text config.  


### Aspect 1

> Idea 1

I like this one. Comes down to using `localhost` as a default when not explicitly provided. We can display it in the UX (read only + refer user to the documentation that would explain it and provide it as an env var if they want).

It gives good defaults that should suffice in simple setups, without keeping the user in the dark.

> That line of argument does make sense, it would also imply doing the same for e.g. announce addresses since they are dependent on the reverse proxy/DNS setup and not really fedimint-defined.

This one is a bit different in nature than location, since the "announce address" is kind of immutable borderline consensus constant, so we don't expect it to be changing. It's more of a public commitment then an internal system implementation detail. So I think it's OK to have it be a part of the config that gets generated in the UI by the operator.

### Aspect 2

Right now we *both* inject `BitcoindRpc` instance *and* pass Bitcoin Rpc connection string to the constructor of `Wallet` module (`Wallet::new_with_bitcoind`) which is the root of my confusion with aspect 2.

So we have to decide:

* either the wallet module is responsible for implementing it's connectivity to Bitcoin Core and we pass only the connection string,
* or the modules can define their dependencies as `trait`/`interface` and in the future we were generalize it as some "Providers"

I was in favour of injecting interface instances and moving towards Providers, but I've realized yesterday that in that case we can't have `IBitcoindRpc` be the interface as it is right now (listing calls we can make the node). While possibly multiple modules will want to talk to the Bitcoin Node, each of them will want to call different methods, and we can't really design up-front a generic interface that would satisfy any module, including calls that bitcoind might add in the future.

We can either have interface that is "you can make any bitcoind jsonrpc call", but then it is not going to be interchangable with Electrum (or is it?).

I hope I explained it well: If we are to use interfaces that modules could share, they need to be clearly defined and generic enough to make future-proof reuse possible, yet concrete enough that we can actually have them implemented.


Given that, I think the idea of re-using interfaces might not be very practical. So maybe we should just relay on textual argument to modules and not bother with interfaces and providers, and we want electrum/esplora, build it into wallet module as a confirable option.


### To sum up

If you don't dissagree the course of action would be:

* Stop passing `BitcoinRpc` interface to the Wallet. We can still use the interface in the wallet module, but just not make it a part of fedimintd-module API boundary. Scrap this PR.
* Pass to each module a subset of environment variables that start from `FM_{MODULE}_` as `env: HashMap<OsString, OsString>` and let them figure out their network connectivity. (Your Idea 2)
* Have wallet module handle both `FM_WALLET_BTC_RPC` and `FM_WALLET_ELECTRUM_RPC`, so we can implement tutorials that have users runs `docker -e FM_WALLET_ELECTRUM_RPC=https://electrum.blockstream.info run` etc. (which was my initial goal).

**Comment by justinmoon** [comment]
>  If not we tell the user in the setup UI to either setup bitcoind or ...

This really begs the question of how the UI will eventually be modularized.

**Comment by dpc** [comment]
> This really begs the question of how the UI will eventually be modularized.

We can have modules (in a generic way) defined a list of keys that they want user to provide values for + functions to validate them. The UI would display them, have user provide values, validate and so on. We could even have the returned list contain type hints for numbers, dates, strings etc. too for better UX.

Quite a bit of extra work, but fundamentally doable, I think.

**Comment by justinmoon** [comment]
We also need to stop hard-coding bitcoind rpc username and password. That bit me and @m1sterc001guy yesterday during our testing. When I passed `username:password@hostname:port` as `bitcoin_rpc` parameter to `run_dkg` I ended up with the following `local.json` config:

```
{
  ...
  "modules": {
    ...
    "wallet": {
      "btc_rpc_address": "username:password@hostname:port",
      "btc_rpc_pass": "bitcoin",
      "btc_rpc_user": "bitcoin"
    }
  }
}
```

**Comment by elsirion** [comment]
>     * Stop passing `BitcoinRpc` interface to the Wallet constructor. We can still use the interface in the wallet module, but just not make it a part of fedimintd-module API boundary.

We probably want to keep a the possibility of injecting around for testing so we don't need to have a textual cfg option for the mock impl, but that shouldn't be a huge problem.

>     * Pass to each module a subset of environment variables that start from `FM_{MODULE}_` as `env: HashMap<OsString, OsString>` and let them figure out their network connectivity. (Your Idea 2)

I think I agree, was wondering if directly reading env vars would be much worse.

> @elsirion Let me know if it seems OK.

SGTM.

**Comment by dpc** [comment]
> I think I agree, was wondering if directly reading env vars would be much worse.

I feel like we should (eventually) not pass all env vars to the modules. If we ever get to the point where we have wasm modules, etc. - in part to sandbox them a bit - 

When thinking about https://github.com/fedimint/fedimint/issues/1231#issuecomment-1374741749 it
